sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldHaveCorrectSignature.cs

```diff
@@ -44,83 +43,116 @@ public sealed class TestMethodShouldHaveCorrectSignature : SonarDiagnosticAnalyz
-                    if (classSymbol == null)
-                    {
-                        return;
-                    }
-
-                    var allFaultyMethods = classSymbol.GetMembers()
-                        .OfType<IMethodSymbol>()
-                        .Select(m => new { method = m, testType = ToKnownTestType(m) })
-                        .Where(tuple => tuple.testType != null)
-                        .Select(
-                            tuple =>
-                            new
-                            {
-                                Location = tuple.method.Locations.First(),
-                                Message = GetFaults(tuple.method, tuple.testType).ToSentence()
-                            })
-                        .Where(tuple => tuple.Message != null);
-
-                    foreach (var faultyMethod in allFaultyMethods)
-                    {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, faultyMethod.Location,
-                            faultyMethod.Message));
-                    }
-                },
-                SyntaxKind.ClassDeclaration);
+            var methodSymbol = c.SemanticModel.GetDeclaredSymbol(c.Node) as IMethodSymbol;
+            if (methodSymbol == null)
+            {
+                return;
+            }
+
+            var validator = GetValidator(methodSymbol);
+            var message = validator(methodSymbol);
+            if (message != null)
+            {
+                c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, methodSymbol.Locations.First(), message));
+            }
         }
 
-        private static KnownType ToKnownTestType(IMethodSymbol method)
+        private static SignatureValidator GetValidator(IMethodSymbol method)
         {
-            return method.GetAttributes()
-                .Select(
-                    attribute =>
-                    {
-                        if (attribute.AttributeClass.Is(KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_TestMethodAttribute))
-                        {
-                            return KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_TestMethodAttribute;
-                        }
-                        else if (attribute.AttributeClass.Is(KnownType.NUnit_Framework_TestAttribute))
-                        {
-                            return KnownType.NUnit_Framework_TestAttribute;
-                        }
-                        else if (attribute.AttributeClass.Is(KnownType.Xunit_FactAttribute))
-                        {
-                            return KnownType.Xunit_FactAttribute;
-                        }
-                        else
-                        {
-                            return null;
-                        }
-                    })
-                .FirstOrDefault();
+            // Find the first matching attribute type in the table
+            var type = method.FindTestAttribute();
```

Comment:
The variable name is not really explicit, I had to look at the return type of the `FindTestAttribute` method to understand. Maybe something like `attributeKnownType`